lightbox: Prevent hero animation between message lists

lakshya1goel opened the pull request:
Fixes #930.
Fixes #44.

## Videos

### Before

https://github.com/user-attachments/assets/37a832c2-87d6-4223-a47e-f5549fccfead


### After

https://github.com/user-attachments/assets/206af9a4-f487-49ad-adca-54faa70d5783

gnprice commented:
Thanks! Chat thread here:
https://chat.zulip.org/#narrow/channel/516-mobile-dev-help/topic/Hero.20animation/near/2092791
with comments particularly on how to write a good test for this fix.

lakshya1goel commented:
Hi @PIG208, PR is ready for a review now, PTAL, Thanks!

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -558,4 +601,51 @@ void main() {
+    testWidgets('no hero animation occurs between different message list pages for same image', (tester) async {
+      final channel = eg.stream(streamId: eg.defaultStreamMessageStreamId);
+      final message = eg.streamMessage(stream: channel,
+        contentMarkdown: ContentExample.spoilerHeaderHasImage.html, topic: 'test topic');
```
nit: This example seems a bit too atypical. How about using something like `imageSingle` to replicate a message that a user is more likely to encounter.

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -558,4 +601,51 @@ void main() {
+      await tester.pump(const Duration(milliseconds: 150));
+
+      final imageInTransition = tester.getRect(imageFinder);
+      check(imageInTransition.top != initialImageRect.top).isTrue();
```
nit: utilize checks API when possible:
```suggestion
      check(imageInTransition.top).not((it) => it.equals(initialImageRect.top));
```
same below

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -558,4 +601,51 @@ void main() {
+      await tester.pumpAndSettle();
+    });
+  });
+}
```
Need a newline at the end of the file.

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -558,4 +601,51 @@ void main() {
+        of: find.byType(StreamMessageRecipientHeader),
+        matching: find.text('test topic')));
+      await tester.pump();
+      await tester.pump(const Duration(milliseconds: 150));
+
+      final imageInTransition = tester.getRect(imageFinder);
+      check(imageInTransition.top).equals(initialImageRect.top);
+      check(imageInTransition.left).equals(initialImageRect.left);
+      await tester.pumpAndSettle();
```
Is this supposed to be mid transition time? While this works, I feel that it can miss bugs if the duration somehow changes.

We can have a while-loop that checks `tester.hasRunningAnimations`, which pumps repeatedly with a fixed duration, and repeats the check:

```suggestion

      int timeElapsed = 0;
      const interval = 50;
      while (timeElapsed < interval || tester.hasRunningAnimations) {
        final imageInTransition = tester.getRect(imageFinder);
        check(imageInTransition.top).equals(initialImageRect.top);
        check(imageInTransition.left).equals(initialImageRect.left);
        await tester.pump(const Duration(milliseconds: interval));
        timeElapsed += interval;
      }
      check(timeElapsed).isGreaterOrEqual(interval);
```

PIG208 replied:
Just saw the [CZO comment](https://chat.zulip.org/#narrow/channel/516-mobile-dev-help/topic/Hero.20animation/near/2095334)[#mobile-dev-help > Hero animation @ 💬](https://chat.zulip.org/#narrow/channel/516-mobile-dev-help/topic/Hero.20animation/near/2095334):

> In order to have that benefit, it's important to keep the two test cases similar to each other as far as possible. It should be easy for the reader to convince themself that the two tests are checking the same thing, and just expecting opposite outcomes.

I agree that having these two tests similar to each other does mitigate the concern of one of them not working properly. So it's probably also fine to leave them as-is. Either way, we should comment that the duration is specifically chosen such that we are in the middle of a hero animation (if there is one).

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,37 @@ class FakeVideoPlayerPlatform extends Fake
+  late PerAccountStore store;
+  late FakeApiConnection connection;
+
+  Future<void> setupMessageListPage(WidgetTester tester, {
```
Let's keep this helper close to where it is used in the 'LightBoxHero' group.

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,37 @@ class FakeVideoPlayerPlatform extends Fake
+
+  Future<void> setupMessageListPage(WidgetTester tester, {
+    Narrow narrow = const CombinedFeedNarrow(),
+    List<Message>? messages,
+    List<ZulipStream>? streams,
```
Because none of the two callers actually rely on having different messages/channels, it looks like we can remove these parameters. An advantage of keeping the helper close to where it's used is that we can more easily spot what properties the tests rely on.

For things that are constant to the tests (like the topic name, finders), we can have a shared local variable within the group.

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -558,4 +601,51 @@ void main() {
+      await setupMessageListPage(tester, narrow: const CombinedFeedNarrow(),
+        streams: [channel], messages: [message]);
+
+      connection.prepare(json:
+        eg.newestGetMessagesResult(foundOldest: true, messages: [message]).toJson());
```
nit: move this to the line just before `tester.tap`, because it prepares an API response for that

PIG208 reviewed (commented):
Thanks for fixing this! The approach looks fine to me. Left a comment on the implementation and some more on the tests.

lakshya1goel commented on lib/widgets/lightbox.dart:
```diff
@@ -44,16 +46,18 @@ class LightboxHero extends StatelessWidget {
   Widget build(BuildContext context) {
     return Hero(
-      tag: _LightboxHeroTag(messageId: message.id, src: src),
+      tag: _LightboxHeroTag(messageId: message.id, src: src, pageContext: pageContext),
```
Things would go wrong since using PageRoot.contextOf(context) here would break the hero animation because the widget exists in two different page contexts simultaneously in the `MessageListPage` and the `lightbox`. By explicitly passing the source page's context (pageContext), we ensure both the source and destination heroes use the same context value in their tags, allowing Flutter to properly match them for the animation.

PIG208 replied:
Indeed. Thanks for the explanation.

lakshya1goel commented on test/widgets/lightbox_test.dart:
```diff
@@ -207,14 +245,19 @@ void main() {
-      // the lightbox route. The lightbox page works together with the route;
-      // it takes the route's entrance animation.
-      await tester.pumpWidget(const ZulipApp());
+      // ZulipApp instead of TestZulipApp because we need:
+      // 1. The navigator to push the lightbox route. The lightbox page works
+      //    together with the route; it takes the route's entrance animation.
+      // 2. The PageRoot widget to provide context for Hero animations between
+      //    the message list and lightbox.
+      await tester.pumpWidget(PageRoot(
+        child: const ZulipApp()
+      ));
       await tester.pump();
       final navigator = await ZulipApp.navigator;
       unawaited(navigator.push(getImageLightboxRoute(
         accountId: eg.selfAccount.id,
+        pageContext: PageRoot.contextOf(navigator.context),
```
No, I think we can't skip passing `pageContext` because during the hero animation, the transitioning widget lives in Flutter's overlay layer outside our PageRoot hierarchy. The hero animation needs the same tag value at both ends (source and destination) to know which widgets to animate between, and since we can't get the PageRoot context during the transition, we must pass it explicitly.

lakshya1goel commented:
Pushed the revision, PTAL @PIG208. Thanks!

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -558,4 +571,77 @@ void main() {
+    late FakeApiConnection connection;
+
+    final channel = eg.stream();
+    final message = eg.streamMessage(stream: channel,
+        contentMarkdown: ContentExample.imageSingle.html, topic: 'test topic');
```
nit: indentation
```suggestion
    final message = eg.streamMessage(stream: channel,
      contentMarkdown: ContentExample.imageSingle.html, topic: 'test topic');
```

PIG208 commented on test/widgets/lightbox_test.dart:
```diff
@@ -558,4 +571,77 @@ void main() {
+        streams: [channel], subscriptions: [subscription]));
+      store = await testBinding.globalStore.perAccount(eg.selfAccount.id);
+      connection = store.connection as FakeApiConnection;
+
+      await store.addUser(eg.selfUser);
```
nit:
```suggestion
      await store.addUser(eg.selfUser);
```
because `store.addUser` is a part of the setup code stanza

PIG208 commented on lib/widgets/content.dart:
```diff
@@ -663,12 +665,14 @@ class MessageImage extends StatelessWidget {
           originalWidth: node.originalWidth,
-          originalHeight: node.originalHeight));
+          originalHeight: node.originalHeight,
+          pageContext: pageContext));
```
While testing this, I tried passing `context` to `pageContext` (and below), and it appears to work. Let's try this so that we don't need to rely on having a PageRoot.

PIG208 reviewed (commented):
Thanks for the update! I found a place where we can potentially shake off the dependency on PageRoot. Let me know what you think!

lakshya1goel commented:
Pushed the revision @PIG208, please have a look. Thanks!

PIG208 commented:
Thanks! I think because we are not using the context of `PageRoot`, there is probably a better name than `pageConext`. (Maybe `messageImageContext`?) Marking this for Greg's review.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -558,4 +566,76 @@ void main() {
+      // Pump halfway through the hero animation duration (300ms)
+      await tester.pump(const Duration(milliseconds: 150));
+
+      final imageInTransition = tester.getRect(imageFinder);
+      check(imageInTransition).top.equals(initialImageRect.top);
+      check(imageInTransition).left.equals(initialImageRect.left);
```
Given `imageFinder = find.byType(RealmContentNetworkImage).first`, this doesn't make an entirely convincing check that there isn't a hero animation going on. This effectively says that the original image is still the *first* RealmContentNetworkImage in the tree — but if there were a hero animation going on, there's no reason that necessarily has to be the first image in the tree. Maybe it comes later in the tree than the original image.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -558,4 +566,76 @@ void main() {
+      // Pump halfway through the hero animation duration (300ms)
+      await tester.pump(const Duration(milliseconds: 150));
+
+      final imageInTransition = tester.getRect(imageFinder);
+      check(imageInTransition).top.not((it) => it.equals(initialImageRect.top));
+      check(imageInTransition).left.not((it) => it.equals(initialImageRect.left));
```
Similarly, this says that the first RealmContentNetworkImage in the tree isn't at the place the original image was. But
* there's the new route, the lightbox, which has a version of the image — maybe that's the first one in the tree now
* there's the old route, the message list, which may have a version of the image — and although right now that won't have moved, maybe in the future we'll use a different navigation transition here so that the old route moves away while the new one is moving in. Again maybe that's the first one in the tree now.

So this should get more specific about our expectations in order to be convincing that it wouldn't end up passing even in some future where we've broken the hero animation entirely.

gnprice reviewed (commented):
Thanks @lakshya1goel for taking this on, and @PIG208 for the previous reviews! Comments below.

lakshya1goel commented on lib/widgets/lightbox.dart:
```diff
@@ -21,20 +21,26 @@ import 'store.dart';
 //   fly to an image preview with a different URL, following a message edit
 //   while the lightbox was open.
```
I think the current implementation handles this scenario well. When a message is edited while its image is open in the lightbox, the behavior depends on whether the image URL changes:
- If the edit changes the image URL:
The Hero animation won't find a matching tag (because src is part of the tag and has changed)
It will gracefully fall back to a fade transition instead of attempting to animate to the wrong image
- If the edit doesn't change the image URL:
All three tag components (messageId, src, and messageImageContext) still match
The Hero animation will work normally

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -300,6 +308,100 @@ void main() {
+      // Pump halfway through the hero animation duration (300ms)
+      await tester.pump(const Duration(milliseconds: 150));
+
+      final imageInTransition = tester.getRect(imageFinder);
+      check(imageInTransition).top.equals(initialImageRect.top);
+      check(imageInTransition).left.equals(initialImageRect.left);
```
This test is also less specific than the previous revision. In fact this one no longer tests anything about the hero animation at all: even if you edit the code so that the tags are completely defeated and the hero animation happens all the time:
```diff
   bool operator ==(Object other) {
-    return other is _LightboxHeroTag &&
-      other.messageId == messageId &&
-      other.src == src &&
-      other.messageImageContext == messageImageContext;
+    return true;
   }
 
   @override
-  int get hashCode => Object.hash('_LightboxHeroTag', messageId, src, messageImageContext);
+  int get hashCode => '_LightboxHeroTag'.hashCode;
```

this test still passes.

Like with the other test case, can you describe what condition this verifies is true, in terms that reflect what the user sees on the screen?

gnprice reviewed (commented):
Thanks for the revision — comments below.

For detailed how-to discussion, let's use [#mobile-dev-help > Hero animation](https://chat.zulip.org/#narrow/channel/516-mobile-dev-help/topic/Hero.20animation/with/2117577), since Zulip will be better for discussion than GitHub is.

gnprice commented:
Thanks! Chat thread here about the tests: [#mobile-dev-help > Hero animation @ 💬](https://chat.zulip.org/#narrow/channel/516-mobile-dev-help/topic/Hero.20animation/near/2131953)

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,152 @@ class FakeVideoPlayerPlatform extends Fake
+      final steps = 50;
+      final stepDuration = heroAnimationDuration ~/ steps;
+
+      for (int i = 1; i <= steps; i++) {
+        await tester.pump(stepDuration);
+
+        final sourceHeroElement = tester.elementList(find.byType(Hero))
+        .where((element) => (element.widget as Hero).tag == heroWidgetTag)
+        .where((element) => element.findAncestorWidgetOfExactType<InteractiveViewer>() == null);
+
+        final destinationHeroElement = tester.elementList(find.byType(Hero))
+        .where((element) => (element.widget as Hero).tag == heroWidgetTag)
+        .where((element) => element.findAncestorWidgetOfExactType<InteractiveViewer>() != null);
+
+        check(sourceHeroElement).isNotEmpty();
+        //without hero animation, destination hero element should not exist
+        check(destinationHeroElement).isEmpty();
```
And more fundamentally: this test does not succeed at doing its job as a test.

The purpose of a test is so that we don't accidentally revert your change. If we do revert your change, the test should fail.

Here:
* Revert your change, but keep the new tests, with the command `git checkout @~ lib/`.
* A different test below needs to be tweaked so it still compiles: comment out the `messageImageContext: navigator.context,` in a call to `getImageLightboxRoute`.
* Run this test file: `flutter test test/widgets/lightbox_test.dart`.
* This test "no hero animation occurs between different message list pages for same image" still passes, even though in reality a hero animation now *does* occur between such pages.
* (And the other test "Hero animation occurs smoothly when opening lightbox from message list" also gets the wrong answer: that one fails, even though in reality the thing it's testing for still works. That's because of that test's use of the `key` that was artificially added for the test — which is a good illustration of why that's not a good way for that test to work.)

gnprice reviewed (commented):
Thanks for the revision. This is closer, but there are still significant things that need to be changed before we can merge it.

In particular, this version still doesn't have a test that successfully exercises the issue this PR is meant to fix. See the last pair of comments below.

lakshya1goel commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,152 @@ class FakeVideoPlayerPlatform extends Fake
+
+      await tester.pumpAndSettle();
+
+      final lightBoxImageFinder = find.byKey(Key(message.id.toString()));
```
actually this was done because when I was finding the image using the src url it is finding two images , this could be due to generation of  source and destination hero image during hero animation , how could I uniquely find out the image element without key ?

lakshya1goel commented:
Hi @gnprice, thanks for the detailed reviews so far. 
I have revised the implementation as:
- Removed the hero related implementation and focused purely on the image now. (Used that earlier to check that the animation is actually a hero animation, nevertheless). 
- Removed the key used for the identification of the image rather used the image src now. 
- Moreover, sorry for the 2nd test, now I have checked and verified that it is failing on reverting my changes and passing after the fix. 
- Resolved the other comments mentioned above.

PTAL and let me know if anything else is required.
Thanks!

gnprice commented on lib/widgets/lightbox.dart:
```diff
@@ -21,39 +21,55 @@ import 'store.dart';
 /// Builds a [Hero] from an image in the message list to the lightbox page.
 class LightboxHero extends StatelessWidget {
   const LightboxHero({
     super.key,
     required this.message,
+    required this.messageImageContext,
     required this.src,
     required this.child,
   });
 
   final Message message;
+  final BuildContext messageImageContext;
   final Uri src;
   final Widget child;
 
   @override
   Widget build(BuildContext context) {
     return Hero(
-      tag: _LightboxHeroTag(messageId: message.id, src: src),
+      tag: _LightboxHeroTag(messageImageContext: messageImageContext, src: src),
```
Is the `message` field on this class used for anything after this change?

lakshya1goel replied:
yeah, its not required now.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +203,136 @@ class FakeVideoPlayerPlatform extends Fake
+      for (int i = 1; i <= steps; i++) {
+        await tester.pump(stepDuration);
+
+        final imageFinder = find.byWidgetPredicate((widget) =>
+          widget is RealmContentNetworkImage && widget.src == imageSrcUrl
+        ).first;
```
Why `.first`?

The purpose of this test is to check that there isn't an image animating around. What if there's one copy that's animating around, and another that's staying in the initial position? The one that stays in the initial position could very well happen to be the first one found in the tree, and then this test would wrongly pass.

lakshya1goel replied:
Here I am getting 2 exactly same images, I have checked the position of both the images inside the loop for 150 steps using the logs and the position is same for both the images for a particular step.
This is probably due to the navigation, on the press of back button the top most route is being popped and we are moving to the older route so at a particular instance we are having the image due to both of these routes.

gnprice replied:
Yeah, that sounds like the behavior we expect to see here.

The job of a test is to convince a skeptical reader that the behavior we expect to see is the behavior that's actually happening (given that the test passes). So in particular this test should be written so that it rules out the possibility I described above, and other alternative possibilities of that flavor.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +203,136 @@ class FakeVideoPlayerPlatform extends Fake
+
+        previousPosition = position;
+      }
+      await tester.pumpAndSettle();
```
If this is needed, that suggests that the animation isn't actually over when this test stops checking.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +203,136 @@ class FakeVideoPlayerPlatform extends Fake
+        initialImagePosition.left - animatedPositions.last.left).abs();
+
+      Rect previousPosition = initialImagePosition;
+      for (final position in animatedPositions) {
+        final positionChangeTop = position.top - previousPosition.top;
+        final positionChangeLeft = position.left - previousPosition.left;
+        check(positionChangeTop.abs()).isLessThan(0.1 * totalTopDistance);
+        check(positionChangeLeft.abs()).isLessThan(0.1 * totalLeftDistance);
```
10% of the distance traveled in the whole animation is quite a lot for a duration of 1/150th of the total duration.

We can measure this better by using the Pythagorean theorem to compute the overall distance moved in the two-dimensional plane, rather than looking at motion in each axis separately. That way if the start and end points happen to be close in one axis, this doesn't expect unreasonably tight bounds on how much the image might move in that axis during the animation.

gnprice replied:
Instead of checking each of these separately in a loop against the same upper bound, it's better to compute the max across all the steps and then check the comparison of just that max against the upper bound.

That will have the same behavior as to whether the test passes or fails. But if it fails, it will give more informative output: it'll print the number from the one step where the image moved the most, instead of just the one that happened to come first out of the steps that were over the bound. That's especially helpful when picking what the bound should be.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +203,136 @@ class FakeVideoPlayerPlatform extends Fake
+        final animatedFlightImageFinder = find.byWidgetPredicate((widget) =>
+          widget is RealmContentNetworkImage && widget.src == imageSrcUrl
+        );
+        final animatedFlightImagePosition = tester.getRect(animatedFlightImageFinder);
+        animatedPositions.add(animatedFlightImagePosition);
```
This can be simpler and shorter by inlining.

gnprice reviewed (commented):
Thanks for the revision. This is significantly closer; some comments below.

gnprice commented:
Thanks for the revision. This looks good now, except that the test needs to be more specific as described in this subthread above: https://github.com/zulip/zulip-flutter/pull/1348#discussion_r2019569770

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,146 @@ class FakeVideoPlayerPlatform extends Fake
+
+        for (final element in currentFrameElements) {
+          final position = tester.getRect(find.byElementPredicate((e) => e == element));
+          imagePositionsOverTime.putIfAbsent(element, () => []);
+          imagePositionsOverTime[element]!.add(position);
+        }
+      }
+
+      for (final element in imagePositionsOverTime.keys) {
+        final positions = imagePositionsOverTime[element]!;
+        check(positions).isNotEmpty();
+
+        final initialPos = positions[0];
+        for (final position in positions) {
+          check(position).equals(initialPos);
```
This still isn't very specific, because it doesn't check that the set of elements at one step has any relationship to the set of elements at another step.

For example, one plausible buggy behavior would be that at each step, there's a new image element and the old image element is gone. From the user's perspective this could produce an animation, because the new element could be at a different position along an animation path. But that would pass this test, because each list would have just one entry in it.

gnprice replied:
Instead, try this:
* Before navigating to the second page, find the one image element and remember its location.
* While on the second page, find the one image element that's now visible and remember its location.
* Once the animation starts, at each step, check that the set of image elements is exactly those two image elements.
* Also at each step, check that each image element is in the same place as it was at the start.

Does that pass? If so, great — that's very specific and makes clear what the behavior is.

If not, then loosen the conditions a little in order to make it pass. But experiment to find just how and why the specific version fails, so that you can loosen it in only the minimal way that makes it pass, keeping it as specific as possible.

lakshya1goel commented:
Hi, @gnprice just pushed the revision following the comments above. The test is passing without loosening the constraints.
PTAL, Thanks!

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,153 @@ class FakeVideoPlayerPlatform extends Fake
+
+        final currentImages = find.byWidgetPredicate((widget) =>
+          widget is RealmContentNetworkImage && widget.src == imageSrcUrl).evaluate();
+        check(currentImages)
+          ..length.equals(2)
+          ..contains(firstElement)
+          ..contains(secondElement);
```
Fun shorthand for this:
```suggestion
        check(currentImages).unorderedEquals([firstElement, secondElement]);
```

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,153 @@ class FakeVideoPlayerPlatform extends Fake
+      for (int i = 0; i < steps; i++) {
+        await tester.pump(stepDuration);
+
+        final currentImages = find.byWidgetPredicate((widget) =>
+          widget is RealmContentNetworkImage && widget.src == imageSrcUrl).evaluate();
```
This is the same finder as `secondImageFinder`, right? It's really "find images with this URL". So it'd be good to combine them, under a name that reflects that shared meaning.

They can also replace `firstImageFinder`, if we just identify the URL in advance.

And we can do that, because this test controls what the URL is going to be — it set up the message list and supplied the message HTML.

Probably cleanest is to just duplicate the URL from that ContentExample value. It's unlikely to change; if it does change, the tests will fail because they won't find elements they expect. So like:
```dart
    final message = eg.streamMessage(stream: channel,
      contentMarkdown: ContentExample.imageSingle.html, topic: 'test topic');
    
    // From ContentExample.imageSingle.
    final imageSrcUrlStr = '/user_uploads/2/ce/nvoNL2LaZOciwGZ-FYagddtK/image.jpg';
    final imageSrcUrl = …;
    final imageFinder = …;
```

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,153 @@ class FakeVideoPlayerPlatform extends Fake
+      final secondImagePosition = tester.getRect(secondImageFinder);
+      final secondElement = secondImageFinder.evaluate().single;
+
+      final backButton = find.byType(BackButton);
+      await tester.tap(backButton);
```
nit: simpler if inlined

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,153 @@ class FakeVideoPlayerPlatform extends Fake
+      final firstImagePosition = tester.getRect(firstImageFinder);
+      final firstElement = firstImageFinder.evaluate().single;
+
+      connection.prepare(json:
+        eg.newestGetMessagesResult(foundOldest: true, messages: [message]).toJson());
+
+      await tester.tap(find.descendant(
+        of: find.byType(StreamMessageRecipientHeader),
+        matching: find.text('test topic')));
+      await tester.pumpAndSettle();
```
nit: these are logically one operation, so join as one stanza:
```suggestion
      connection.prepare(json:
        eg.newestGetMessagesResult(foundOldest: true, messages: [message]).toJson());
      await tester.tap(find.descendant(
        of: find.byType(StreamMessageRecipientHeader),
        matching: find.text('test topic')));
      await tester.pumpAndSettle();
```

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,153 @@ class FakeVideoPlayerPlatform extends Fake
+      const heroAnimationDuration = Duration(milliseconds: 300);
+      const steps = 150;
+      final stepDuration = heroAnimationDuration ~/ steps;
+      List<Rect> animatedPositions = [];
```
nit: final

gnprice replied:
`const` (which is stronger than `final`) was fine for the locals where it applied — this comment was meant for `animatedPositions` :slightly_smiling_face: 

(In GitHub's review interface, if you make a comment targeting just a single line — like this one was — it shows the 4 lines up through that one as the context. That does makes it look a bit ambiguous, though in cases where the difference matters it is possible to find if you look closely: some comments say "Comment on lines MMM to NNN" above them.)

The principle here is that when a value is marked as immutable, that makes the code easier for the reader to reason about. So whenever we don't actually need to mutate it (like we do for loop variables like `previousPosition`), it's best to declare it `final` or `const`.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,153 @@ class FakeVideoPlayerPlatform extends Fake
+    late FakeApiConnection connection;
+
+    final channel = eg.stream();
+    final message = eg.streamMessage(stream: channel,
+      contentMarkdown: ContentExample.imageSingle.html, topic: 'test topic');
```
nit: order elements logically: so stream, topic, then content, because the stream and topic give the more general context where the content appears

lakshya1goel commented:
Hi, @gnprice I have pushed the revision. PTAL, Thanks!

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,128 @@ class FakeVideoPlayerPlatform extends Fake
+    late FakeApiConnection connection;
+
+    final channel = eg.stream();
+    final imageSrcUrlStr = 'https://chat.example/user_uploads/thumbnail/2/ce/nvoNL2LaZOciwGZ-FYagddtK/image.jpg/840x560.webp';
```
nit: this should have a comment saying where it comes from, as it otherwise looks quite random — cf https://github.com/zulip/zulip-flutter/pull/1348#discussion_r2029628420

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,128 @@ class FakeVideoPlayerPlatform extends Fake
+
+      await setupMessageListPage(tester);
+
+      final messageContentFinder = find.byWidgetPredicate((widget) =>
+        widget is MessageContent && widget.message.id == message.id
+      );
+      final messageListImageFinder = find.descendant(
+        of: messageContentFinder,
+        matching: find.byType(RealmContentNetworkImage)
+      );
+      final initialImagePosition = tester.getRect(messageListImageFinder);
```
This can be simplified to work the same way as the corresponding part of the other test case. (And then `imageFinder` can be shared between the test cases.)

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,128 @@ class FakeVideoPlayerPlatform extends Fake
+        final currentImages = imageFinder.evaluate();
+        check(currentImages).unorderedEquals([firstElement, secondElement]);
+
+        check(tester.getRect(
+          find.byElementPredicate((e) => e == firstElement))).equals(firstImagePosition);
```
nit: line too long; instead:
```suggestion
        check(tester.getRect(find.byElementPredicate((e) => e == firstElement)))
          .equals(firstImagePosition);
```

gnprice replied:
Better yet, make a little helper function within this test case, and then:
```suggestion
        check(getElementRect(firstElement)).equals(firstImagePosition);
```

Then the same `getElementRect` can be used for defining `firstImagePosition` in terms of `firstElement` in the first place. That way this check works the same way as `firstImagePosition` got computed originally — so we're doing the same thing the same way both times, instead of two somewhat different ways, and that means one fewer thing for the reader to understand and think through.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,128 @@ class FakeVideoPlayerPlatform extends Fake
+      for (int i = 0; i < steps; i++) {
+        await tester.pump(stepDuration);
+
+        final currentImages = imageFinder.evaluate();
+        check(currentImages).unorderedEquals([firstElement, secondElement]);
```
nit: simpler if inlined

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,128 @@ class FakeVideoPlayerPlatform extends Fake
+        widget is RealmContentNetworkImage && widget.src == imageSrcUrl
+      );
+      final firstImagePosition = tester.getRect(imageFinder);
+      final firstElement = imageFinder.evaluate().single;
```
Bit cleaner (as `evaluate` is meant really as internal API for implementing these finders):
```suggestion
      final firstElement = tester.element(imageFinder);
```

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,128 @@ class FakeVideoPlayerPlatform extends Fake
+      Rect previousPosition = initialImagePosition;
+      double maxStepDistance = 0.0;
+      for (final position in animatedPositions) {
+        final stepDistance = sqrt(
+          pow(position.top - previousPosition.top, 2) +
+          pow(position.left - previousPosition.left, 2));
```
The logic here can become a bit easier to see by making a small helper function inside the test case:
```suggestion
        final stepDistance = dist(previousPosition, position);
```

Particularly for comparison with `totalDistance` above.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,128 @@ class FakeVideoPlayerPlatform extends Fake
+      store = await testBinding.globalStore.perAccount(eg.selfAccount.id);
+      connection = store.connection as FakeApiConnection;
+      await store.addUser(eg.selfUser);
+
+      connection.prepare(json:
+        eg.newestGetMessagesResult(foundOldest: true, messages: [message]).toJson());
+
+      await tester.pumpWidget(TestZulipApp(accountId: eg.selfAccount.id,
+        child: MessageListPage(initNarrow: const CombinedFeedNarrow())));
+
+      await tester.pumpAndSettle();
```
nit: these are basically one operation; join them as one stanza:
```suggestion

      connection.prepare(json:
        eg.newestGetMessagesResult(foundOldest: true, messages: [message]).toJson());
      await tester.pumpWidget(TestZulipApp(accountId: eg.selfAccount.id,
        child: MessageListPage(initNarrow: const CombinedFeedNarrow())));
      await tester.pumpAndSettle();
```

gnprice reviewed (commented):
Thanks for the revision! This is getting closer; here's comments on this version.

(This review is a bit delayed because I was out sick part of last week.)

lakshya1goel commented:
Hi, I have pushed the revision. PTAL @gnprice. Thanks!

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,112 @@ class FakeVideoPlayerPlatform extends Fake
+      final stepDuration = heroAnimationDuration ~/ steps;
+      for (int i = 0; i < steps; i++) {
+        await tester.pump(stepDuration);
+        check(imageFinder.evaluate()).unorderedEquals([firstElement, secondElement]);
```
nit: similar comment as: https://github.com/zulip/zulip-flutter/pull/1348#discussion_r2045447107

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,112 @@ class FakeVideoPlayerPlatform extends Fake
+    late PerAccountStore store;
+    late FakeApiConnection connection;
+
+    final channel = eg.stream();
+    // From ContentExample.imageSingle.
+    final imageSrcUrlStr = 'https://chat.example/user_uploads/thumbnail/2/ce/nvoNL2LaZOciwGZ-FYagddtK/image.jpg/840x560.webp';
+    final imageSrcUrl = Uri.parse(imageSrcUrlStr);
+    final message = eg.streamMessage(stream: channel,
+      topic: 'test topic', contentMarkdown: ContentExample.imageSingle.html);
+
+    final imageFinder = find.byWidgetPredicate(
+      (widget) => widget is RealmContentNetworkImage && widget.src == imageSrcUrl);
```
nit: this is clearer if rearranged (like at https://github.com/zulip/zulip-flutter/pull/1348#discussion_r2029628420):
```suggestion
    final channel = eg.stream();
    final message = eg.streamMessage(stream: channel,
      topic: 'test topic', contentMarkdown: ContentExample.imageSingle.html);

    // From ContentExample.imageSingle.
    final imageSrcUrlStr = 'https://chat.example/user_uploads/thumbnail/2/ce/nvoNL2LaZOciwGZ-FYagddtK/image.jpg/840x560.webp';
    final imageSrcUrl = Uri.parse(imageSrcUrlStr);
    final imageFinder = find.byWidgetPredicate(
      (widget) => widget is RealmContentNetworkImage && widget.src == imageSrcUrl);
```

There's basically two narratives (or mini-narratives) here: setting up `message`, and setting up `imageFinder`. This way the reader gets through one, then goes through the other, rather than interleaving them.

This way also makes the comment on `imageSrcUrlStr` make more sense — the reason `ContentExample.imageSingle` is relevant is that it's what's used in `message`. When this comes shortly after `message` rather than before it, the reader has that context.

gnprice reviewed (commented):
Thanks for the revision! Down now to a couple of small comments, both of them echoes of comments from the last round.

gnprice commented on test/widgets/lightbox_test.dart:
```diff
@@ -197,6 +204,115 @@ class FakeVideoPlayerPlatform extends Fake
+      final stepDuration = heroAnimationDuration ~/ steps;
+      for (int i = 0; i < steps; i++) {
+        await tester.pump(stepDuration);
+        check(tester.element(imageFinder.first)).anyOf([
+          (e) => e.equals(firstElement), (e) => e.equals(secondElement)]);
+        check(tester.element(imageFinder.last)).anyOf([
+          (e) => e.equals(firstElement), (e) => e.equals(secondElement)]);
```
This no longer checks that the situation is the way we expect: it allows there to be three or more elements matching `imageFinder`.

It's also a lot more code to read than an `unorderedEquals` check. (Cf. my previous comment https://github.com/zulip/zulip-flutter/pull/1348#discussion_r2029614529 .)

Using `unorderedEquals` is the right way to do this. You'll want something different from `tester.element`, though.

Take a look around the API surrounding `tester.element` to see what variations of it you can find that would be useful here. I recommend jumping in your IDE to the definition of `tester.element` — that will let you browse around the neighboring code.

gnprice commented on pubspec.lock:
```diff
@@ -650,26 +650,26 @@ packages:
     dependency: transitive
     description:
       name: leak_tracker
-      sha256: "6bb818ecbdffe216e81182c2f0714a2e62b593f4a4f13098713ff1685dfb6ab0"
+      sha256: "8dcda04c3fc16c14f48a7bb586d4be1f0d1572731b6d81d51772ef47c02081e0"
```
This diff doesn't seem related to what this commit is doing. To keep the commits [clear and coherent](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html), this change should be kept to a separate commit.

It should also go with a corresponding change in `pubspec.yaml`. See our README about upgrading Flutter.

lakshya1goel commented:
Hi @gnprice, I have pushed the correction please have a look. Thanks!

gnprice reviewed (commented):
Thanks @lakshya1goel for all your work on this!

The tests all look good now, modulo one nit.

See also one more-substantive comment below. But I think it doesn't call for any code changes, only comments and metadata.

So I'll make those edits and merge. I'll also add a commit that revises the docs from this PR.

From reading the TODO comment that I remarked on in an earlier review, I see that this in fact fixes not only #930 but also another issue #44. A good followup for you would be to do next would be to write a regression test to ensure that #44 stays fixed; see details below.
